// vnpay/lib/enums/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VnpTransactionType = exports.VnpOrderType = exports.VnpLocale = exports.VnpCurrCode = exports.UrlService = void 0;
var UrlService;
(function (UrlService) {
    UrlService["sandbox"] = "https://sandbox.vnpayment.vn/paymentv2/vpcpay.html";
})(UrlService || (exports.UrlService = UrlService = {}));
var VnpCurrCode;
(function (VnpCurrCode) {
    VnpCurrCode["VND"] = "VND";
})(VnpCurrCode || (exports.VnpCurrCode = VnpCurrCode = {}));
var VnpLocale;
(function (VnpLocale) {
    VnpLocale["VN"] = "vn";
    VnpLocale["EN"] = "en";
})(VnpLocale || (exports.VnpLocale = VnpLocale = {}));
var VnpOrderType;
(function (VnpOrderType) {
    VnpOrderType["OTHER"] = "other";
})(VnpOrderType || (exports.VnpOrderType = VnpOrderType = {}));
var VnpTransactionType;
(function (VnpTransactionType) {
    VnpTransactionType["PAYMENT"] = "01";
    VnpTransactionType["FULL_REFUND"] = "02";
    VnpTransactionType["PARTIAL_REFUND"] = "03";
})(VnpTransactionType || (exports.VnpTransactionType = VnpTransactionType = {}));
