// vnpay/lib/enums/product-code.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductCode = void 0;
/**
 * @see https://sandbox.vnpayment.vn/apis/docs/loai-hang-hoa/
 */
var ProductCode;
(function (ProductCode) {
	ProductCode["Food_Consumption"] = "100000";
	ProductCode["Phone_Tablet"] = "110000";
	ProductCode["ElectricAppliance"] = "120000";
	ProductCode["Computers_OfficeEquipment"] = "130000";
	ProductCode["Electronics_Sound"] = "140000";
	ProductCode["Books_Newspapers_Magazines"] = "150000";
	ProductCode["Sports_Picnics"] = "160000";
	ProductCode["Hotel_Tourism"] = "170000";
	ProductCode["Cuisine"] = "180000";
	ProductCode["Entertainment_Training"] = "190000";
	ProductCode["Fashion"] = "200000";
	ProductCode["Health_Beauty"] = "210000";
	ProductCode["Mother_Baby"] = "220000";
	ProductCode["KitchenUtensils"] = "230000";
	ProductCode["Vehicle"] = "240000";
	ProductCode["Pay"] = "250000";
	ProductCode["AirlineTickets"] = "250007";
	ProductCode["CardCode"] = "260000";
	ProductCode["Pharmacy_MedicalServices"] = "270000";
	ProductCode["Other"] = "other";
})(ProductCode || (exports.ProductCode = ProductCode = {}));
